src/providers/akamai/mod.rs

```diff
@@ -0,0 +1,283 @@
+use std::collections::HashMap;
+
+use crate::providers::MetadataProvider;
+use crate::retry;
+
+const TOKEN_URL: &'static str = "http://169.254.169.254/v1/token";
+const INSTANCE_METADATA_URL: &'static str = "http://169.254.169.254/v1/instance";
+const NETWORK_METADATA_URL: &'static str = "http://169.254.169.254/v1/network";
+const SSHKEYS_URL: &'static str = "http://169.254.169.254/v1/ssh-keys";
+
+static HDR_TOKEN_EXPIRY: &str = "metadata-token-expiry-seconds";
+static HDR_TOKEN: &str = "metadata-token";
+
+static TOKEN_TTL: &str = "3600";
+
+pub struct AkamaiProvider {
+    client: retry::Client,
+}
+
+impl AkamaiProvider {
+    /// Instantiate a new `AkamaiProvider`.
+    pub fn try_new() -> Result<Self> {
+        let client = retry::Client::try_new()?;
+        Ok(Self { client })
+    }
+
+    fn token(&self) -> Result<HeaderValue> {
+        let token: String = self
+            .client
+            .put(retry::Raw, TOKEN_URL.to_string(), None)
+            .header(
+                HeaderName::from_static(HDR_TOKEN_EXPIRY),
+                HeaderValue::from_static(TOKEN_TTL),
+            )
+            .dispatch_put()?
+            .ok_or_else(|| anyhow!("get metadata token"))?;
+
+        let token = HeaderValue::from_str(&token).context("create header value from token")?;
+        Ok(token)
+    }
+
+    /// Fetch the instance metadata.
+    fn fetch_instance_metadata(&self) -> Result<Instance> {
+        let token = self.token()?;
+        let instance: Instance = self
+            .client
+            .get(retry::Json, INSTANCE_METADATA_URL.to_string())
+            .header(HeaderName::from_static(HDR_TOKEN), token)
+            .header(
+                HeaderName::from_static("accept"),
+                HeaderValue::from_static("application/json"),
+            )
+            .send()?
+            .ok_or_else(|| anyhow!("instance metadata not found"))?;
+        Ok(instance)
+    }
+
+    /// Fetch the network metadata.
+    fn fetch_network_metadata(&self) -> Result<Network> {
+        let network: Network = self
+            .client
+            .get(retry::Json, NETWORK_METADATA_URL.to_string())
+            .header(HeaderName::from_static(HDR_TOKEN), self.token()?)
+            .header(
+                HeaderName::from_static("accept"),
+                HeaderValue::from_static("application/json"),
+            )
+            .send()?
+            .ok_or_else(|| anyhow!("network metadata not found"))?;
+        Ok(network)
+    }
+
+    /// Fetch the SSH keys.
+    fn fetch_ssh_keys(&self) -> Result<SshKeys> {
+        let ssh_keys: SshKeys = self
+            .client
+            .get(retry::Json, SSHKEYS_URL.to_string())
+            .header(HeaderName::from_static(HDR_TOKEN), self.token()?)
+            .header(
+                HeaderName::from_static("accept"),
+                HeaderValue::from_static("application/json"),
+            )
+            .send()?
+            .ok_or_else(|| anyhow!("ssh keys not found"))?;
```

Comment:
If no SSH keys are provided through the Metadata service, the result is: 

```json
{"users": {}}
```

so I think we would need the `SshKeys` struct around, at the very least, for deserialization.

Instead of returning the `SshKeys` struct, would you prefer to have `fetch_ssh_keys` return a `HashMap<String, Vec<String>>`?

Reply:
No, that's fine. Just wanted to make sure we were handling the no-SSH-keys case correctly.

Reply:
Okay. I opted to use a `HashMap<String, Vec<String>>` anyways, since it was fewer custom types to carry around after deserialization.